patches/tModLoader/Terraria/ModLoader/ItemLoader.cs

```diff
@@ -153,19 +144,33 @@ private class HookList
 			if (IsModItem(item.type) && createModItem)
 				item.ModItem = GetItem(item.type).Clone(item);
 
-			item.globalItems = InstancedGlobals.Select(g => g.Clone(item, item)).ToArray();
+			item.globalItems = globalItems
+				.Where(g => g.InstanceForEntity(item, false))
+				.Select(g => new Instanced<GlobalItem>(g.index, g.InstancePerEntity ? g.Clone(item, item) : g))
+				.ToArray();
 
 			item.ModItem?.AutoDefaults();
 			item.ModItem?.SetDefaults();
 
-			foreach (var g in HookSetDefaults.arr)
-				g.Instance(item).SetDefaults(item);
+			//Could potentially be sped up.
+			var lateInitGlobals = globalItems
+				.Where(g => !item.globalItems.Any(i => i.index == g.index) && g.InstanceForEntity(item, true))
+				.Select(g => new Instanced<GlobalItem>(g.index, g.InstancePerEntity ? g.Clone(item, item) : g));
+
+			item.globalItems = item.globalItems
+				.Union(lateInitGlobals)
+				.OrderBy(i => i.index)
+				.ToArray();
```

Comment:
Move to common, say.. `LoaderUtils`? Either 2 methods, or one with a lambda passed?